src/stream/src/executor/backfill.rs

```diff
@@ -447,6 +570,125 @@ where
+        if let Some(current_pos_inner) = current_pos {
+            // state w/o vnodes.
+            Self::build_temporary_state(current_state, is_finished, current_pos_inner);
+            Self::flush_data(table, epoch, old_state, current_state).await?;
+            *old_state = Some(current_state.into());
+        } else {
+            table.commit_no_data_expected(epoch);
+        }
+        Ok(())
+    }
+
+    /// Flush the data
+    async fn flush_data(
+        table: &mut StateTable<S>,
+        epoch: EpochPair,
+        old_state: &mut Option<Vec<Datum>>,
+        current_partial_state: &mut [Datum],
+    ) -> StreamExecutorResult<()> {
+        let vnodes = table.vnodes().clone();
+        if let Some(old_state) = old_state {
+            if old_state[1..] == current_partial_state[1..] {
+                table.commit_no_data_expected(epoch);
+                return Ok(());
+            } else {
+                vnodes.iter_ones().for_each(|vnode| {
+                    let datum = Some((vnode as i16).into());
+                    current_partial_state[0] = datum.clone();
+                    old_state[0] = datum;
+                    table.write_record(Record::Update {
+                        old_row: &old_state[..],
+                        new_row: &(*current_partial_state),
+                    })
+                });
+            }
+        } else {
+            // No existing state, create a new entry.
+            vnodes.iter_ones().for_each(|vnode| {
+                let datum = Some((vnode as i16).into());
+                // fill the state
+                current_partial_state[0] = datum;
+                table.write_record(Record::Insert {
+                    new_row: &(*current_partial_state),
+                })
+            });
+        }
+        table.commit(epoch).await
+    }
+
+    // We want to avoid building a row for every vnode.
+    // Instead we can just modify a single row, and dispatch it to state table to write.
+    fn build_temporary_state(row_state: &mut [Datum], is_finished: bool, current_pos: &OwnedRow) {
+        row_state[1..current_pos.len() + 1].clone_from_slice(current_pos.as_inner());
+        row_state[current_pos.len() + 1] = Some(is_finished.into());
+    }
+
+    fn update_pos(chunk: &StreamChunk, pk_in_output_indices: &[usize]) -> Option<OwnedRow> {
+        Some(
+            chunk
+                .rows()
+                .last()
+                .unwrap()
+                .1
+                .project(pk_in_output_indices)
+                .into_owned_row(),
+        )
+    }
+
+    // TODO(kwannoel): I'm not sure if ["None" ..] encoding is appropriate
+    // for the case where upstream snapshot is empty, and we want to persist
+    // backfill state as "finished".
+    // Could it be confused with another case where pk position was indeed empty?
+    // I don't think it will because they both record that backfill is finished.
+    // We can revisit in future if necessary.
```

Comment:
Please take note regarding this case. I'm not sure if it can have any implications.

Reply:
Using `None` looks good to me.